Add a new blog post: Plan for Dune 3.0

snowleopard opened the pull request:
A blog post to discuss Dune 3.0 with the community. 

[Rendered preview](https://github.com/ocaml/dune-www/blob/a984b59609a9d81f13e2964696f6d7bb3d9e2f1d/_posts/2021-02-26-dune-3-plan.md)

TODO: After publishing the post, create a Discuss page, and link to it.

emillon commented on _posts/2021-02-26-dune-3-plan.md:
```diff
@@ -0,0 +1,87 @@
+The Dune team has started planning the next major release and we would like to
+share and discuss the current plan with the community. We expect to release
+Dune 3 in August 2021. As with previous major releases, we will put a lot of
+effort in backward compatibility, which means the majority of Dune users should
```
I think that's always a good opportunity to explain the dune version vs dune-lang version - upgrading dune itself is still a safe operation as it will not change how existing projects are built. Some project-related breaking changes will happen (e.g. scoping semantics) but these will only happen when you as a maintainer flip the switch with `(lang dune 3.0)`.

snowleopard replied:
Thanks! I agree, it's good to remind about the difference. However, in the specific case of scoping semantics, we do plan to change the behaviour unconditionally, because otherwise it would be really difficult to reason about nested projects that specify different version of the Dune language. It seems worth clarifying this in the text.

avsm replied:
Absolutely -- otherwise there would be no point doing a major version bump if there was no way to break compat.

It might be worth having a short section explicitly about forward and backwards compatibility -- just today on the opam meeting it emerged that the opam developers didn't know about the "dune-lang" _vs_ "dune" distinction and were working around a dune bug as a result.  The post about dune 3 is a good chance to set expectations: users can continue to use dune-lang in dune 3.0, _except_ for the planned breaking changes around scoping (which are important to continue to scale the monorepo approach).

snowleopard replied:
I've updated the paragraph in the intro, trying to keep it brief. Hope it makes things clearer.

It seems useful to have a separate page going into more details about the Dune vs Dune language distinction, which we can link to from various places (including such blog posts).

snowleopard replied:
I made another pass and a separate section on breaking changes. Please have a look.

emillon reviewed (commented):
This reads very well. Thanks for doing this!

avsm commented on _posts/2021-02-26-dune-3-plan.md:
```diff
@@ -0,0 +1,87 @@
+  it. This will be a breaking change but nested vendored projects are not very
+  common, so this change will only affect a small number of users and in a
+  positive way.
+
```
The scoping semantics are the only breaking change from a dune-lang perspective and may be worth pulling out specially to explain the whole monorepo situation. The others could all (in theory) be put into an incremental release of dune 2.x.

snowleopard replied:
I'm a bit hesitant to put the emphasis on scoping semantics as the only breaking change. The change in warnings, for example, might break things too (if a script runs `dune` and expects empty output). Similar for some other changes.

The new intro says that some changes cannot be guarded by the Dune language version, and I think scoping semantics is one such change (but not the only one).

jeremiedimino replied:
> The change in warnings, for example, might break things too (if a script runs dune and expects empty output). Similar for some other changes.

This kind of breaking changes are less dramatic. Dune should be able to work on a workspace composed of different projects developed by different people and using different versions of the language. This is why backward compatibility at the language level is very important. For the rest, such as the cli behaviour, this is still important but we can allow ourselves a bit more freedom to change things at each major release. The only cli invocation we should be particularly mindful about is `dune build -p <package>`, because that's the one we tell people to use in their opam file.

snowleopard replied:
I added a separate section on breaking changes with respect to the language, adding one more change that we previously overlooked. Please have a look.

avsm commented on _posts/2021-02-26-dune-3-plan.md:
```diff
@@ -0,0 +1,87 @@
+
+* **Support for OPAM v1.** We believe that it's not used any more, so we'd like
+  to remove the associated code.
+
```
You can link to https://opam.ocaml.org/blog/deprecating-opam-1-2-0/ if useful -- it has been EOLed for some years now.

snowleopard replied:
Thanks for the link, added!

avsm commented on _posts/2021-02-26-dune-3-plan.md:
```diff
@@ -0,0 +1,87 @@
+  of the Dune language, which changes over time. It is therefore a good practice
+  to create a `dune-project` file to freeze the semantics of the `dune` file to a specific version of the language. Dune used to provide some support for creating and updating `dune-project` files automatically but in practice this feature caused a fair amount of confusion for users, so we are removing it.
+
+* **Calling ocamlfind at runtime.** When running Dune outside of an OPAM
```
minor: it's `opam` and not `OPAM`

snowleopard replied:
Fixed.

snowleopard commented:
@emillon @avsm We're planning to publish the blog post soon, so it would be great to hear if the new version addresses your concerns about the way we describe breaking changes.

avsm commented:
belatedly, it comprehensively addressed all my comments @snowleopard :-)

snowleopard commented:
@avsm Cool, thanks :-)
